### PR TITLE
vote and timeout handlers

### PR DESCRIPTION
**Pull request by wjrjerome**

### CODE REVIEW

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 1]
```diff
@@ -326,10 +371,19 @@ func (x *XDPoS_v2) verifyVotingRule(header *types.Header) error {
 }
 
 // Once Hot stuff voting rule has verified, this node can then send vote
-func (x *XDPoS_v2) sendVote(header *types.Header) error {
+func (x *XDPoS_v2) sendVote(blockInfo utils.BlockInfo) error {
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go]: `*utils.BlockInfo` ? (Not sure, let's ask @liam-lai )

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 2]
```diff
@@ -110,29 +99,27 @@ func (x *XDPoS_v2) VerifyHeader(chain consensus.ChainReader, header *types.Heade
...
-	// 2. Only 1 message processing at the time
-	return nil
-}
-
 /*
 	SyncInfo workflow
 */
-// Verify syncInfo and trigger trigger process QC or TC if successful
+// Verify syncInfo and trigger process QC or TC if successful
 func (x *XDPoS_v2) VerifySyncInfoMessage(syncInfo utils.SyncInfo) error {
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go:106]: `*utils. SyncInfo ` ? (Not sure, let's ask @liam-lai )

File: tests/consensus/v2_test.go [hunk 3]
```diff
@@ -24,3 +26,243 @@ func TestCountdownTimeoutToSendTimeoutMessage(t *testing.T) {
...
+	// Set round to 1
+	engineV2.SetNewRoundFaker(utils.Round(1), false)
+	err = engineV2.TimeoutHandler(timeoutMsg)
+	assert.NotNil(t, err)
+	// Timeout msg round < currentRound
+	assert.Equal(t, "Timeout message round number: 2 does not match currentRound: 1", err.Error())
+}
+
+// VoteHandler
+func TestVoteMessageHandlerSuccessfullyGenerateTCandSyncInfo(t *testing.T) {
```
Comment (wgr523) [inline tests/consensus/v2_test.go]: SuccessfullyGenerateQC (no SyncInfo)

File: tests/consensus/v2_test.go [hunk 4]
```diff
@@ -24,3 +26,243 @@ func TestCountdownTimeoutToSendTimeoutMessage(t *testing.T) {
...
+	assert.Equal(t, utils.Round(1), engineV2.GetCurrentRound())
+	voteMsg = &utils.Vote{
+		ProposedBlockInfo: *blockInfo,
+		Signature:         []byte{2},
+	}
+	err = engineV2.VoteHandler(*voteMsg)
+	assert.Nil(t, err)
+	assert.Equal(t, utils.Round(1), engineV2.GetCurrentRound())
+
+	// Create a vote message that should trigger vite pool hook
```
Comment (wgr523) [inline tests/consensus/v2_test.go]: vite -> vote

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 5]
```diff
@@ -376,9 +459,18 @@ func (x *XDPoS_v2) onCountdownTimeout(time time.Time) error {
...
 }
 
 func (x *XDPoS_v2) getCurrentRoundMasterNodes() []common.Address {
 	return []common.Address{}
 }
+
+func (x *XDPoS_v2) getSyncInfo() utils.SyncInfo {
+	return utils.SyncInfo{
+		HighestQuorumCert:  *x.highestQuorumCert,
+		HighestTimeoutCert: *x.highestTimeoutCert,
```
Comment (wgr523) [inline consensus/XDPoS/engines/engine_v2/engine.go]: what if they are `nil`?

**Review by wgr523 — APPROVED** [review]

File: consensus/XDPoS/engines/engine_v2/engine.go (hunk 5 above)
Comment (wjrjerome) [reply consensus/XDPoS/engines/engine_v2/engine.go]: they should not be nil. they are initilised with 
```highestTimeoutCert: &utils.TimeoutCert{},
		highestQuorumCert:  &utils.QuorumCert{},
```
in the `New` method

File: consensus/XDPoS/engines/engine_v2/engine.go (hunk 2 above)
Comment (wjrjerome) [reply consensus/XDPoS/engines/engine_v2/engine.go:106]: we can talk about it later. We need to make sure all verify function have the same pattern.

File: consensus/XDPoS/engines/engine_v2/engine.go [hunk 6]
```diff
@@ -376,9 +459,18 @@ func (x *XDPoS_v2) onCountdownTimeout(time time.Time) error {
...
+		x.BroadcastCh <- msg
+	}()
 }
 
 func (x *XDPoS_v2) getCurrentRoundMasterNodes() []common.Address {
 	return []common.Address{}
 }
+
+func (x *XDPoS_v2) getSyncInfo() utils.SyncInfo {
+	return utils.SyncInfo{
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/engine.go:472]: could we change utils.SyncInfo struct from reference type to pointer type, easier for other functions to pass and initiate. 
```
	HighestQuorumCert  QuorumCert
	HighestTimeoutCert TimeoutCert
```
```
	HighestQuorumCert  *QuorumCert
	HighestTimeoutCert *TimeoutCert
```

File: tests/consensus/v2_test.go [hunk 7]
```diff
@@ -24,3 +26,243 @@ func TestCountdownTimeoutToSendTimeoutMessage(t *testing.T) {
...
+
+	tc := syncInfoMsg.(utils.SyncInfo).HighestTimeoutCert
+	assert.NotNil(t, tc)
+	assert.Equal(t, tc.Round, utils.Round(1))
+	sigatures := []utils.Signature{[]byte{1}, []byte{2}, []byte{3}}
+	assert.ElementsMatch(t, tc.Signatures, sigatures)
+	assert.Equal(t, utils.Round(2), engineV2.GetCurrentRound())
+}
+
+func TestThrowErrorIfTimeoutMsgRoundLessThanCurrentRound(t *testing.T) {
```
Comment (liam-lai) [inline tests/consensus/v2_test.go]: LessThan -> NotEqual

File: tests/consensus/v2_test.go [hunk 8]
```diff
@@ -24,3 +26,243 @@ func TestCountdownTimeoutToSendTimeoutMessage(t *testing.T) {
...
 }
+
+// Timeout handler
+func TestTimeoutMessageHandlerSuccessfullyGenerateTCandSyncInfo(t *testing.T) {
+	blockchain, _, _, _ := PrepareXDCTestBlockChain(t, 11, params.TestXDPoSMockChainConfigWithV2Engine)
+	engineV2 := blockchain.Engine().(*XDPoS.XDPoS).EngineV2
+
+	// Set round to 1
+	engineV2.SetNewRoundFaker(utils.Round(1), false)
+	// Create two timeout message which will not reach timeout pool threshold
```
Comment (liam-lai) [inline tests/consensus/v2_test.go:37]: Doesn't need this cmd

File: tests/consensus/v2_test.go [hunk 9]
```diff
@@ -24,3 +26,243 @@ func TestCountdownTimeoutToSendTimeoutMessage(t *testing.T) {
...
+	assert.Equal(t, utils.Round(2), engineV2.GetCurrentRound())
+}
+
+func TestThrowErrorIfTimeoutMsgRoundLessThanCurrentRound(t *testing.T) {
+	blockchain, _, _, _ := PrepareXDCTestBlockChain(t, 11, params.TestXDPoSMockChainConfigWithV2Engine)
+	engineV2 := blockchain.Engine().(*XDPoS.XDPoS).EngineV2
+
+	// Set round to 3
+	engineV2.SetNewRoundFaker(utils.Round(3), false)
+	// Create two timeout message which will not reach timeout pool threshold
```
Comment (liam-lai) [inline tests/consensus/v2_test.go]: Doesn't need this cmd

File: tests/consensus/v2_test.go [hunk 10]
```diff
@@ -24,3 +26,243 @@ func TestCountdownTimeoutToSendTimeoutMessage(t *testing.T) {
...
+
+	blockInfo := &utils.BlockInfo{
+		Hash:   common.HexToHash("0x1"),
+		Round:  utils.Round(2),
+		Number: big.NewInt(999),
+	}
+
+	// Set round to 3
+	engineV2.SetNewRoundFaker(utils.Round(3), false)
+	// Create two timeout message which will not reach timeout pool threshold
```
Comment (liam-lai) [inline tests/consensus/v2_test.go]: Doesn't need this.

File: tests/consensus/v2_test.go [hunk 11]
```diff
@@ -24,3 +26,243 @@ func TestCountdownTimeoutToSendTimeoutMessage(t *testing.T) {
...
+		Signature:         []byte{3},
+	}
+
+	err = engineV2.VoteHandler(*voteMsg)
+	assert.Nil(t, err)
+	// Check round has now changed from 1 to 2
+	assert.Equal(t, utils.Round(2), engineV2.GetCurrentRound())
+}
+
+func TestThrowErrorIfVoteMsgRoundLessThanCurrentRound(t *testing.T) {
```
Comment (liam-lai) [inline tests/consensus/v2_test.go]: LessThan -> NotEqual

**Review by liam-lai — APPROVED** [review]

File: tests/consensus/v2_test.go (hunk 8 above)
Comment (wjrjerome) [reply tests/consensus/v2_test.go:37]: why not? First two will not trigger anything, but the third will

File: tests/consensus/v2_test.go (hunk 11 above)
Comment (wjrjerome) [reply tests/consensus/v2_test.go]: 👍

**Pull request merged**